File: detectorOptimizado.py
import re
import nltk
import time
from concurrent.futures import ThreadPoolExecutor
from nltk.corpus import cess_esp
import os

if not os.path.join(nltk.data.path[0], 'cess_esp'):
    nltk.download('cess_esp')

def distancia_levenshtein(s1, s2, umbral):
    m, n = len(s1), len(s2)
    if m < n:
        return distancia_levenshtein(s2, s1, umbral)

    previous_row = range(n + 1)
    for i, c1 in enumerate(s1):
        current_row = [i + 1]
        for j, c2 in enumerate(s2):
            insertions = previous_row[j + 1] + 1
            deletions = current_row[j] + 1
            substitutions = previous_row[j] + (c1 != c2)
            current_row.append(min(insertions, deletions, substitutions))
        if min(current_row) > umbral:
            return umbral + 1  # Si la mínima distancia supera el umbral, no es necesario continuar
        previous_row = current_row

    return previous_row[-1]

def es_palabra_incorrecta(palabra, diccionario, umbral=0.5):
    if palabra in diccionario:
        return False
    palabra_len = len(palabra)
    with ThreadPoolExecutor() as executor:
        resultados = list(executor.map(lambda palabra_correcta: abs(len(palabra_correcta) - palabra_len) > umbral or
                                                          distancia_levenshtein(palabra, palabra_correcta, umbral) > umbral, diccionario))
    return all(resultados)

def limpiar_texto(texto):
    texto = texto.lower()
    texto = re.sub(r'[^\w\s]', '', texto)
    return texto

def contar_palabras_incorrectas(texto, diccionario, umbral=0):
    texto_limpio = limpiar_texto(texto)
    palabras = texto_limpio.split()
    palabras_incorrectas = 0
    detalles_incorrectos = []

    for palabra in palabras:
        if es_palabra_incorrecta(palabra, diccionario, umbral):
            palabras_incorrectas += 1
            detalles_incorrectos.append(palabra)

    return palabras_incorrectas, detalles_incorrectos

def cargar_diccionario(archivo):
    with open(archivo, 'r', encoding='utf-8') as file:
        diccionario = set(file.read().splitlines())  # Usar set para búsquedas rápidas
    return diccionario

def cargar_diccionario_nltk():
    palabras = cess_esp.words()
    diccionario = set(palabras)
    return diccionario

# Cargar diccionario desde nltk
#diccionario = cargar_diccionario_nltk()

# Medir el tiempo de ejecución
start_time = time.time()

archivo_diccionario = 'esp3.txt'

textos = [
    "El trbajo eta inompleto poqe fue apido",
    "La combnicación de cocolates y flores fué ideal",
    "Me gustaría comprar un perro de raca pequeña",
    "Estoy deseando probar esa nuva marca de helado",
    "No entendí el acertijo, era demaciado complicado",
    "Mi hermano compitió en una competición de artes marciales",
    "El café estava delicioso, no me gusto el pastel",
    "El estudiante tenía que entregar un travajo muy extenso",
    "Fuimos al cine pero la película estubo aburrida",
    "Esa novela tiene una trama apasionante, la recomiendo",
]

for texto in textos:
    palabras_incorrectas, detalles_incorrectos = contar_palabras_incorrectas(texto, cargar_diccionario_nltk())
    print(f"Texto: {texto}")
    print(f"Número de palabras incorrectas: {palabras_incorrectas}")
    print(f"Palabras incorrectas: {detalles_incorrectos}")
    print()

end_time = time.time()
elapsed_time = end_time - start_time

print(f"Tiempo de ejecución: {elapsed_time:.2f} segundos")

File: detectorOptimizado.java
import java.io.*;
import java.nio.file.*;
import java.util.*;
import java.util.concurrent.*;


public class detectorOptimizado {
    private static final double UMBRAL = 0.5;

    public static void main(String[] args) throws IOException {
        String archivoDiccionario = "esp1.txt";
        Set<String> diccionario = cargarDiccionario(archivoDiccionario);

        String[] textos = {
            "El trbajo eta inompleto poqe fue apido",
            "La combnicación de cocolates y flores fué ideal",
            "Me gustaría comprar un perro de raca pequeña",
            "Estoy deseando probar esa nuva marca de helado",
            "No entendí el acertijo, era demaciado complicado",
            "Mi hermano compitió en una competición de artes marciales",
            "El café estava delicioso, no me gusto el pastel",
            "El estudiante tenía que entregar un travajo muy extenso",
            "Fuimos al cine pero la película estubo aburrida",
            "Esa novela tiene una trama apasionante, la recomiendo",
        };

        long startTime = System.currentTimeMillis();

        for (String texto : textos) {
            Resultado resultado = contarPalabrasIncorrectas(texto, diccionario, UMBRAL);
            System.out.println("Texto: " + texto);
            System.out.println("Número de palabras incorrectas: " + resultado.numeroIncorrectas);
            System.out.println("Palabras incorrectas: " + resultado.palabrasIncorrectas);
            System.out.println();
        }

        long endTime = System.currentTimeMillis();
        long elapsedTime = endTime - startTime;
        System.out.printf("Tiempo de ejecución: %.2f segundos\n", elapsedTime / 1000.0);
    }

    private static Set<String> cargarDiccionario(String archivo) throws IOException {
        return new HashSet<>(Files.readAllLines(Paths.get(archivo)));
    }

    private static class Resultado {
        int numeroIncorrectas;
        List<String> palabrasIncorrectas;

        Resultado(int numeroIncorrectas, List<String> palabrasIncorrectas) {
            this.numeroIncorrectas = numeroIncorrectas;
            this.palabrasIncorrectas = palabrasIncorrectas;
        }
    }

    private static Resultado contarPalabrasIncorrectas(String texto, Set<String> diccionario, double umbral) {
        String textoLimpio = limpiarTexto(texto);
        String[] palabras = textoLimpio.split("\\s+");
        int palabrasIncorrectas = 0;
        List<String> detallesIncorrectos = new ArrayList<>();

        for (String palabra : palabras) {
            if (esPalabraIncorrecta(palabra, diccionario, umbral)) {
                palabrasIncorrectas++;
                detallesIncorrectos.add(palabra);
            }
        }

        return new Resultado(palabrasIncorrectas, detallesIncorrectos);
    }

    private static boolean esPalabraIncorrecta(String palabra, Set<String> diccionario, double umbral) {
        if (diccionario.contains(palabra)) {
            return false;
        }
        int palabraLen = palabra.length();
        ExecutorService executor = Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors());
        List<Future<Boolean>> resultados = new ArrayList<>();

        for (String palabraCorrecta : diccionario) {
            resultados.add(executor.submit(() -> {
                if (Math.abs(palabraCorrecta.length() - palabraLen) > umbral * palabraLen) {
                    return true;
                }
                return distanciaLevenshtein(palabra, palabraCorrecta, (int) (umbral * palabraLen)) > umbral * palabraLen;
            }));
        }

        executor.shutdown();
        try {
            for (Future<Boolean> resultado : resultados) {
                if (!resultado.get()) {
                    return true;
                }
            }
        } catch (InterruptedException | ExecutionException e) {
            e.printStackTrace();
        }

        return false;
    }

    private static int distanciaLevenshtein(String s1, String s2, int umbral) {
        int m = s1.length();
        int n = s2.length();
        if (m < n) {
            return distanciaLevenshtein(s2, s1, umbral);
        }

        int[] previousRow = new int[n + 1];
        int[] currentRow = new int[n + 1];

        for (int j = 0; j <= n; j++) {
            previousRow[j] = j;
        }

        for (int i = 1; i <= m; i++) {
            currentRow[0] = i;
            for (int j = 1; j <= n; j++) {
                int insertions = previousRow[j] + 1;
                int deletions = currentRow[j - 1] + 1;
                int substitutions = previousRow[j - 1] + (s1.charAt(i - 1) != s2.charAt(j - 1) ? 1 : 0);
                currentRow[j] = Math.min(insertions, Math.min(deletions, substitutions));
            }
            if (currentRow[minIndex(currentRow)] > umbral) {
                return umbral + 1;
            }
            int[] temp = previousRow;
            previousRow = currentRow;
            currentRow = temp;
        }

        return previousRow[n];
    }

    private static int minIndex(int[] array) {
        int minIndex = 0;
        for (int i = 1; i < array.length; i++) {
            if (array[i] < array[minIndex]) {
                minIndex = i;
            }
        }
        return minIndex;
    }

    private static String limpiarTexto(String texto) {
        return texto.toLowerCase().replaceAll("[^\\w\\s]", "");
    }
}
